### PR TITLE
cmd/bosun: only enable miniprofiler if there's a human

### LINKED ISSUE #938: Memory leak in Bosun
I updated our test servers to the latest version of bosun from https://github.com/bosun-monitor/bosun/releases/download/20150428222252/bosun-linux-amd64
After running for slightly less than a day, it stopped responding.

The command line where I started it revealed:
```
 ./bosun-linux-amd64 -c=/data/bosun.conf
2015/05/04 16:21:54 enabling syslog
Killed
```
Syslog (cat /var/log/messages |grep bosun) did not reveal any log messages in the hours before the crash.

It looks like a memory leak. The graph of bosun.collect.alloc grew gradually from 200Mb after deploying the new version to 12Gb just before the "crash":
rapid memory leak

Looking back over the last week at the memory behaviour of the previous version, there was a similar memory growth pattern in the previous version too but at a much slower rate. The bottom graph shows gradual memory increasing over the course of a week followed by two rapid increases for the newer version. 
bosun memory leak memory only last 7 days

Just for interest sake, here is a general Bosun dashboard; the other stats look reasonable. Although there is a high number of go routines after restarting Bosun this appears unrelated to the leak.
bosun memory leak dashboard

More information about our setup:
* Backend: OpenTSDB
* Data is being passed through Bosun to OpenTSDB (as visible from the dashboard)
* We send data points every minute at a rate of about 37000 per minute
* In addition scollector is submitting data from one machine monitoring openTSDB, elasticsearch, Bosun, Linux and os
* The rule file is still a small prototype:

```
httpListen = :8070
tsdbHost = localhost:4242

smtpHost = ******
emailFrom = ******

macro grafanaConfig {
    $grafanaHost = ******
}

notification emailIzak {
	email = izak.marais@vastech.co.za
	next = emailIzak
	timeout = 24h
}


##################### Templates #######################


template generic {
    body = `{{template "genericHeader" .}}
    {{template "genericDef" .}}
    
    {{template "genericTags" .}}

    {{template "genericComputation" .}}
    
     {{if .Alert.Vars.graph}}
     <h3>{{.Alert.Vars.graphTitle}}</h3>
    <p>{{.Graph .Alert.Vars.graph}}
    {{end}}`
    
    subject =  {{.Last.Status}}: {{.Alert.Name}} on instance {{.Group.serviceinstance}}
}

template genericHeader {   
    body = `
    <h3> Possible actions </h3>   
    {{if .Alert.Vars.note}}
        <p>{{.Alert.Vars.note}}
    {{end}}
     <p><a href="{{.Ack}}">Acknowledge alert</a>
   
    {{if .Alert.Vars.grafanaDash}}
        <p><a href="{{.Alert.Vars.grafanaDash}}"> View the relevant statistics dasboard </a>
    {{end}}
    `
}

template genericDef {
    body = `
    <h3> Details </h3>
    <p><strong>Alert definition:</strong>
    <table>
        <tr>
            <td>Name:</td>
            <td>{{replace .Alert.Name "." " " -1}}</td></tr>
        <tr>
            <td>Warn:</td>
            <td>{{.Alert.Warn}}</td></tr>
        <tr>
            <td>Crit:</td>
            <td>{{.Alert.Crit}}</td></tr>
    </table>`
}

template genericTags {
    body = `<p><strong>Tags</strong>
    
    <table>
        {{range $k, $v := .Group}}
            {{if eq $k "host"}}
                <tr><td>{{$k}}</td><td><a href="{{$.HostView $v}}">{{$v}}</a></td></tr>
            {{else}}
                <tr><td>{{$k}}</td><td>{{$v}}</td></tr>
            {{end}}
        {{end}}
    </table>`
}

template genericComputation {
    body = `
    <p><strong>Computation</strong>
    
    <table>
        {{range .Computations}}
            <tr><td><a href="{{$.Expr .Text}}">{{.Text}}</a></td><td>{{.Value}}</td></tr>
        {{end}}
    </table>`
}

template unkown {
	subject = {{.Name}}: {{.Group | len}} unknown alerts. 
	body = `
	<p>Unknown alerts imply no data is being recorded for their monitored time series. Therefore we cannot know what is happening. 
	<p>Time: {{.Time}}
	<p>Name: {{.Name}}
	<p>Alerts:
	{{range .Group}}
		<br>{{.}}
	{{end}}`
}

unknownTemplate = unkown


#################### alerts #######################


alert FlowRouterBytesZero {
	template = generic
	$query = "sum:bytes.bytes.counter.value{serviceinstance=*}"
	
	$note = The flow router has reported zero bytes in the last 2 minutes. This note should contain extra information specifying what action the operator should take to resolve it. 
	$graph =q($query, "24h", "")
	$graphTitle = Flow router traffic in the last 24 hours
	macro = grafanaConfig
	$grafanaDash = $grafanaHost/dashboard/db/per-flow-route-bytes-drill-down
	
	$avgBytesPer2Min = avg(q($query, "2m", ""))
	$avgBytesPer5Min = avg(q($query, "5m", ""))
	
	warn =  $avgBytesPer2Min == 0
	crit =  $avgBytesPer5Min == 0
	critNotification = emailIzak
}

```

### PR DESCRIPTION
**Pull request by mjibson**
fixes #938

### CODE REVIEW

File: cmd/bosun/web/static/js/0-bosun.ts [hunk 1]
```diff
@@ -78,6 +78,21 @@ bosunApp.config(['$routeProvider', '$locationProvider', function($routeProvider:
 		otherwise({
 			redirectTo: '/',
 		});
+	$httpProvider.interceptors.push(function($q) {
```
Comment (captncraig) [inline cmd/bosun/web/static/js/0-bosun.ts]: so this wraps all angular-originated api calls with the miniprofiler flag? What calls will come from a different context that we don't want miniprofiler on?

**Comment by captncraig** [comment]
generate

File: cmd/bosun/web/static/js/0-bosun.ts (hunk 1 above)
Comment (mjibson) [inline cmd/bosun/web/static/js/0-bosun.ts]: The goal here is do enable miniprofiler only if there's a human in front of it. So anything that uses the API (like scollector and other /api/put or /api/metadata/put agents) will not trigger miniprofiler. This was causing out of memory issues with the raintank folks.

**Comment by captncraig** [comment]
LGTM

**Comment by mjibson** [comment]
generated. is rechecking.

**Pull request merged**